### DISCUSSION TITLE
How to integrate a gcc patch in riscv-none-elf-gcc-xpack ?

### DISCUSSION THREAD

**Discussion by bvernoux** [xpack-dev-tools/riscv-none-elf-gcc-xpack#7] [Q&A] [answered]
A contributor David Carne (@davidcarne) have created the patch [gcc-12.2.0-riscv-wch-fast-interrupt.patch](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/files/10258787/gcc-12.2.0-riscv-wch-fast-interrupt.patch) to add support of "WCH-Interrupt-fast" for WCH WCH56x(and potentially other WCH RISC-V MCU)
- This patch has been tested with https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/releases/tag/v12.2.0-1.1

This patch work fine (see the toolchain pre-built with this patch https://hydrabus.com/hydrausb3_toolchain)
So far I have not found how to integrate it (in a clean way) with [riscv-none-elf-gcc-xpack](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack) (My workaround is to patch the Work/cache/gcc-12.2.0.tar.xz with this patch)
Does anyone could help to integrate that patch ?

**Comment by ilg-ul** [comment]
If you want only to integrate the patch into your local build, there is a mechanism to apply a patch after the archive is downloaded. The patch must be located in the `patches` folder, and have a specific name, including the version.

If you would like to have the patch included in this distribution, things are more complicated, you have to contribute the patches to GNU GCC, and after the next GCC is released, they'll be in the xPack too.

**Reply by bvernoux** [reply]
I would like both in fact
1) Short term to have the procedure to put the patch in the paches folder (which one and with which exact name and format ?)
    - To have also a way to rebuild it without using Docker if possible (on a Ubuntu LTS distribution for different target at least Linux & Windows)
2) Mid/Long term to have the patch included in GNU GCC if you can help on that you are welcome

**Comment by ilg-ul** [comment, marked as answer]
Identify the function that downloads and extracts the gcc sources and somewhere there you should identify a variable which holds the patch name. 

Please note that the build scripts will shortly be updated to a new version of the build environment.

For contributing the patch to GNU, I suggest you first discuss the issues with the RISC-V GCC maintainers, I am not involved in development, I'm just compiling the sources.

**Reply by bvernoux** [reply]
Could you help with the exact process for the patch ? (or maybe I shall wait new version of the build environment if it is simpler ?)
- What I understand (from the documentation/guess) is the patch shall be copied in `Work/riscv-none-elf-gcc-xpack.git/scripts/helper/patches/`
  - For information the patch [gcc-12.2.0-riscv-wch-fast-interrupt.patch](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/files/10258787/gcc-12.2.0-riscv-wch-fast-interrupt.patch) to apply on gcc shall be applied with -p1 (so it seems it shall be renamed gcc-12.2.0-riscv-wch-fast-interrupt.patch.diff)
  - Example: 
    ```
    cd ./gcc-12.2.0/
    patch -p1 < gcc-12.2.0-riscv-wch-fast-interrupt.patch
    ```
- It seems the key point will be to use the patch filename here https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/blob/5a06b88c7a06ada6efc5c5c50215a7cb884b272a/scripts/common-versions-source.sh#L565 to be confirmed ...
- So the process will be to copy and rename the patch [gcc-12.2.0-riscv-wch-fast-interrupt.patch](https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack/files/10258787/gcc-12.2.0-riscv-wch-fast-interrupt.patch) to` Work/riscv-none-elf-gcc-xpack.git/scripts/helper/patches/gcc-12.2.0.patch.diff` then just build the toolchain ?

**Reply by ilg-ul** [reply]
For the current version I confirm:

- the `gcc-${GCC_VERSION}.patch.diff` patch name (not very inspired, might be changed).
- the location is the local `patches` folder.

For details, try to find where `GCC_PATCH_FILE_NAME` is used, somewhere there is the `download_and_extract` function called.

**Reply by bvernoux** [reply]
I have tried to build it with your details but it was not "Applying" the patch (the build was fine but without applying the patch on gcc source, checked with source code/deploy files vs my reference patch applied by hand)
```
rm -rf ${HOME}/Work/riscv-none-elf-gcc-xpack.git; \
git clone \
  --branch v12.2.0-1.1 \
  https://github.com/xpack-dev-tools/riscv-none-elf-gcc-xpack.git \
  ${HOME}/Work/riscv-none-elf-gcc-xpack.git; \
git -C ${HOME}/Work/riscv-none-elf-gcc-xpack.git submodule update --init --recursive

cp ${HOME}/gcc-12.2.0-riscv-wch-fast-interrupt.patch ${HOME}/Work/riscv-none-elf-gcc-xpack.git/patches/gcc-12.2.0.patch.diff

bash ${HOME}/Work/riscv-none-elf-gcc-xpack.git/scripts/helper/build.sh preload-images
docker images
docker system prune --force

screen -S arm
sudo rm -rf ~/Work/riscv-none-elf-gcc-*-*
bash ${HOME}/Work/riscv-none-elf-gcc-xpack.git/scripts/helper/build.sh --develop --disable-tests --disable-multilib --linux64 --win64
```

**Reply by ilg-ul** [reply]
Are you sure you have a `patches/gcc-12.2.0.patch.diff` file?

In `common-versions-source.sh` add a `set -x` before calling the `download_cross_gcc` function and an `exit 1` after it, clean everything and rerun. This should give you a better picture of what is happening.

**Reply by bvernoux** [reply]
I confirm I have  `~/Work/riscv-none-elf-gcc-xpack.git/patches/gcc-12.2.0.patch.diff` (file size is 4083 bytes)
I obtain following log:
```
+ echo '[tar -x' -f /Host/home/ben/Work/cache/gcc-12.2.0.tar.xz '--no-same-owner]'
[tar -x -f /Host/home/ben/Work/cache/gcc-12.2.0.tar.xz --no-same-owner]
+ tar -x -f /Host/home/ben/Work/cache/gcc-12.2.0.tar.xz --no-same-owner
+ '[' 3 -ge 3 ']'
+ cd gcc-12.2.0
+ do_patch ''
+ '[' '!' -z '' ']'
+ chmod -R +w gcc-12.2.0
+ local with_update_config_sub=
+ '[' '' == y ']'
+ exit 1
```
Which confirm the do_patch  is applied on empty file instead of expected gcc-12.2.0.patch.diff

I have understood the issue there was a missing 
`GCC_PATCH_FILE_NAME="gcc-${GCC_VERSION}.patch.diff"` just before the 1st download_cross_gcc

**Reply by ilg-ul** [reply]
Oops! Yes, the variable must be defined **before** the function call.

**Reply by bvernoux** [reply]
Thanks for your help
